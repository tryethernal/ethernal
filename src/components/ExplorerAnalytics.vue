<template>
    <v-container fluid>
        <v-card outlined style="height: 100%">
            <v-card-text>
                <v-row>
                    <v-col cols="3" class="pb-0">
                        <v-select @change="initAllCharts" hide-details="true" dense primary outlined label="Time Range" :items="ranges" v-model="selectedTimeRange" item-text="label" item-value="value">
                        </v-select>
                    </v-col>
                </v-row>
                <v-row>
                    <v-col cols="12" md="6">
                        <Line-Chart :title="'Transaction Volume'" :xLabels="charts['transactionVolume'].xLabels" :data="charts['transactionVolume'].data" :tooltipUnit="'transaction'" :index="0" />
                    </v-col>

                    <v-col cols="12" md="6">
                        <Line-Chart :title="'ERC20 Transfer Volume'" :xLabels="charts['erc20TransferVolume'].xLabels" :data="charts['erc20TransferVolume'].data" :tooltipUnit="'transfers'" :index="1" />
                    </v-col>

                    <v-col cols="12" md="6">
                        <Line-Chart :title="'Average Gas Price'" :xLabels="charts['averageGasPrice'].xLabels" :data="charts['averageGasPrice'].data" :tokenSymbol="'gwei'" :floating="true" :index="2" />
                    </v-col>

                    <v-col cols="12" md="6">
                        <Line-Chart :title="'Average Transaction Fee'" :xLabels="charts['averageTransactionFee'].xLabels" :data="charts['averageTransactionFee'].data" :tokenSymbol="chain.token || 'ETH'" :floating="true" :index="3" />
                    </v-col>

                    <v-col cols="12" md="6">
                        <Line-Chart :title="'Active Wallets Count'" :xLabels="charts['uniqueWalletCount'].xLabels" :data="charts['uniqueWalletCount'].data" :tooltipUnit="'wallet'" :index="4" />
                    </v-col>

                    <v-col cols="12" md="6">
                        <Line-Chart :title="'Cumulative Wallets Count'" :xLabels="charts['cumulativeWalletCount'].xLabels" :data="charts['cumulativeWalletCount'].data" :tooltipUnit="'wallet'" :index="5" />
                    </v-col>

                    <v-col cols="12" md="6">
                        <Line-Chart :title="'Deployed Contracts Count'" :xLabels="charts['deployedContractCount'].xLabels" :data="charts['deployedContractCount'].data" :tooltipUnit="'contract'" :index="6" />
                    </v-col>

                    <v-col cols="12" md="6">
                        <Line-Chart :title="'Cumulative Deployed Contracts Count'" :xLabels="charts['cumulativeDeployedContractCount'].xLabels" :data="charts['cumulativeDeployedContractCount'].data" :tooltipUnit="'contract'" :index="6" />
                    </v-col>
                </v-row>
            </v-card-text>
        </v-card>
    </v-container>
</template>

<script>
const moment = require('moment');
const ethers = require('ethers');
import { mapGetters } from 'vuex';
import LineChart from './LineChart';

export default {
    name: 'ExplorerAnalytics',
    components: {
        LineChart
    },
    data: () => ({
        selectedTimeRange: 14,
        ranges: [
            { label: '14 Days', value: 14 },
            { label: '30 Days', value: 30 },
            { label: 'All Time', value: 0 }
        ],
        charts: {
            transactionVolume: {},
            erc20TransferVolume: {},
            averageGasPrice: {},
            averageTransactionFee: {},
            uniqueWalletCount: {},
            cumulativeWalletCount: {},
            deployedContractCount: {},
            cumulativeDeployedContractCount: {}
        },
    }),
    mounted() {
       this.initAllCharts();
    },
    methods: {
        moment,
        initAllCharts() {
            this.getTransactionVolume();
            this.getErc20TransferVolume();
            this.getAverageGasPrice();
            this.getAverageTransactionFee();
            this.getUniqueWalletCount();
            this.getCumulativeWalletCount();
            this.getDeployedContractCount();
            this.getCumulativeDeployedContractCount();
        },
        getTransactionVolume() {
            this.server.getTransactionVolume(this.from, this.to)
                .then(({ data }) => {
                    this.charts['transactionVolume'] = {
                        xLabels: data.map(t => t.date),
                        data: data.map(t => parseInt(t.count))
                    };
                })
                .catch(console.log);
        },
        getErc20TransferVolume() {
            this.server.getTokenTransferVolume(this.from, this.to, null, 'erc20')
                .then(({ data }) => {
                    this.charts['erc20TransferVolume'] = {
                        xLabels: data.map(t => t.date),
                        data: data.map(t => parseInt(t.count))
                    };
                })
                .catch(console.log);
        },
        getAverageGasPrice() {
            this.server.getAverageGasPrice(this.from, this.to)
                .then(({ data }) => {
                    this.charts['averageGasPrice'] = {
                        xLabels: data.map(t => t.date),
                        data: data.map(t => parseFloat(ethers.utils.formatUnits(ethers.BigNumber.from(t.avg), 'gwei')))
                    };
                })
                .catch(console.log);
        },
        getAverageTransactionFee() {
            this.server.getAverageTransactionFee(this.from, this.to)
                .then(({ data }) => {
                    this.charts['averageTransactionFee'] = {
                        xLabels: data.map(t => t.date),
                        data: data.map(t => parseFloat(ethers.utils.formatUnits(ethers.BigNumber.from(t.avg), 'ether')))
                    };
                })
                .catch(console.log);
        },
        getUniqueWalletCount() {
            this.server.getUniqueWalletCount(this.from, this.to)
                .then(({ data }) => {
                    this.charts['uniqueWalletCount'] = {
                        xLabels: data.map(t => t.date),
                        data: data.map(t => parseInt(t.count))
                    };
                })
                .catch(console.log);
        },
        getCumulativeWalletCount() {
            this.server.getCumulativeWalletCount(this.from, this.to)
                .then(({ data }) => {
                    this.charts['cumulativeWalletCount'] = {
                        xLabels: data.map(t => t.date),
                        data: data.map(t => parseInt(t.count))
                    };
                })
                .catch(console.log);
        },
        getDeployedContractCount() {
            this.server.getDeployedContractCount(this.from, this.to)
                .then(({ data }) => {
                    this.charts['deployedContractCount'] = {
                        xLabels: data.map(t => t.date),
                        data: data.map(t => parseInt(t.count))
                    };
                })
                .catch(console.log);
        },
        getCumulativeDeployedContractCount() {
            this.server.getCumulativeDeployedContractCount(this.from, this.to)
                .then(({ data }) => {
                    this.charts['cumulativeDeployedContractCount'] = {
                        xLabels: data.map(t => t.date),
                        data: data.map(t => parseInt(t.count))
                    };
                })
                .catch(console.log);
        }
    },
    computed: {
        ...mapGetters([
            'chain',
        ]),
        from() {
            return this.selectedTimeRange > 0 ? new Date(new Date() - this.selectedTimeRange * 24 * 3600 * 1000) : new Date(0);
        },
        to() {
            return new Date();
        }
    }
}
</script>
