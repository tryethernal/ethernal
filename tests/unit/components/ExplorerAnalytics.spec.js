import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';
import ExplorerAnalytics from '@/components/ExplorerAnalytics.vue';

let helper;
const stubs = [
    'Line-Chart',
];

describe('ExplorerAnalytics.vue', () => {

    beforeEach(() => {
        jest.clearAllMocks();
        helper = new MockHelper();
    });

    it('Should display charts', async () => {
        jest.spyOn(helper.mocks.server, 'getTransactionVolume')
            .mockResolvedValueOnce({ data: [
                { date: 0, count: 2 },
                { date: 1, count: 2 }
            ]});

        jest.spyOn(helper.mocks.server, 'getTokenTransferVolume')
            .mockResolvedValueOnce({ data: [
                { date: 0, count: 2 },
                { date: 1, count: 2 }
            ]});

        jest.spyOn(helper.mocks.server, 'getAverageGasPrice')
            .mockResolvedValueOnce({ data: [
                { date: 0, avg: '200000000000000000' },
                { date: 1, avg: '300000000000000000' }
            ]});

        jest.spyOn(helper.mocks.server, 'getAverageTransactionFee')
            .mockResolvedValueOnce({ data: [
                { date: 0, avg: '200000000000000000' },
                { date: 1, avg: '300000000000000000' }
            ]});

        jest.spyOn(helper.mocks.server, 'getUniqueWalletCount')
            .mockResolvedValueOnce({ data: [
                { date: 0, count: 2 },
                { date: 1, count: 2 }
            ]});

        jest.spyOn(helper.mocks.server, 'getCumulativeWalletCount')
            .mockResolvedValueOnce({ data: [
                { date: 0, count: 2 },
                { date: 1, count: 2 }
            ]});

        jest.spyOn(helper.mocks.server, 'getDeployedContractCount')
            .mockResolvedValueOnce({ data: [
                { date: 0, count: 2 },
                { date: 1, count: 2 }
            ]});

        jest.spyOn(helper.mocks.server, 'getCumulativeDeployedContractCount')
            .mockResolvedValueOnce({ data: [
                { date: 0, count: 2 },
                { date: 1, count: 2 }
            ]});

        const wrapper = helper.mountFn(ExplorerAnalytics, { stubs });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });
});
