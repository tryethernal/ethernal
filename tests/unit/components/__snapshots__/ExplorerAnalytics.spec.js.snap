// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExplorerAnalytics.vue Should display charts 1`] = `
<div class="container container--fluid">
  <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
    <div class="v-card__text">
      <div class="row">
        <div class="pb-0 col col-3">
          <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
            <div class="v-input__control">
              <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-5" class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">â€‹</span></legend>
                </fieldset>
                <div class="v-select__slot"><label for="input-5" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Time Range</label>
                  <div class="v-select__selections">
                    <div class="v-select__selection v-select__selection--comma">14 Days</div><input primary="" id="input-5" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                  </div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                  </div><input type="hidden" value="14">
                </div>
                <div class="v-menu">
                  <!---->
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-md-6 col-12">
          <line-chart-stub tooltipunit="transaction" title="Transaction Volume" index="0" data="2,2" xlabels="0,1"></line-chart-stub>
        </div>
        <div class="col-md-6 col-12">
          <line-chart-stub tooltipunit="transfers" title="ERC20 Transfer Volume" index="1" data="2,2" xlabels="0,1"></line-chart-stub>
        </div>
        <div class="col-md-6 col-12">
          <line-chart-stub floating="true" tokensymbol="gwei" title="Average Gas Price" index="2" data="200000000,300000000" xlabels="0,1"></line-chart-stub>
        </div>
        <div class="col-md-6 col-12">
          <line-chart-stub floating="true" tokensymbol="Ether" title="Average Transaction Fee" index="3" data="0.2,0.3" xlabels="0,1"></line-chart-stub>
        </div>
        <div class="col-md-6 col-12">
          <line-chart-stub tooltipunit="wallet" title="Active Wallets Count" index="4" data="2,2" xlabels="0,1"></line-chart-stub>
        </div>
        <div class="col-md-6 col-12">
          <line-chart-stub tooltipunit="wallet" title="Cumulative Wallets Count" index="5" data="2,2" xlabels="0,1"></line-chart-stub>
        </div>
        <div class="col-md-6 col-12">
          <line-chart-stub tooltipunit="contracts" title="Deployed Contracts Count" index="6" data="2,2" xlabels="0,1"></line-chart-stub>
        </div>
        <div class="col-md-6 col-12">
          <line-chart-stub tooltipunit="contract" title="Cumulative Deployed Contracts Count" index="6" data="2,2" xlabels="0,1"></line-chart-stub>
        </div>
      </div>
    </div>
  </div>
</div>
`;
